src/Callback.php

```diff
@@ -43,6 +43,8 @@ protected function init(): void
     public function setUrlTrigger(string $trigger = null)
     {
         $this->urlTrigger = $trigger ?: $this->name;
+
+        $this->getOwner()->stickyGet($this->urlTrigger);
```

Comment:
Now I understand how this PR is supposed to work - what is a `Callback` definition?

Because if more than one is added, tree/independent design is gone because of shared sticky.

Reply:
A callback is a GET request, that when triggered execute a function. The function should terminate the code execution and render a View (normally the one consuming the callback) via JSON. 
Many Views may consume callback at the same time but only one of them can terminate execution.

Reply:
I am in favor of keeping `mergeStickyArgsFromChildView` as it allows View to control which child View can modify it's sticky args. Without this method (as this PR is coded now), the Callback modifies a parent View, which introduces shared/global state which is not wanted in my opinion.

Reply:
I think `mergeStickyArgsFromChildView` is actually not necessary, or I do not understand the use case there. The callback can be triggered, i.e. execute the set function or terminate, i.e. render the proper View as JSON.
The view is actually controlling their stickyArgs, so nested View with callback can still be triggered.

Reply:
If callback has some args that should be sticky, with `mergeStickyArgsFromChildView` the args are stored inside the Callback object. If that object is mutated to some different state, the parent/independent view can reflect that.

If some URL in callback needs to be generated without the Callback sticky args, with this PR, it is not possible as the args are merged in the parent View together.

With this reasoning, it makes sense to me to keep the sticky args not hardcoded with the parent component, but instead of keep `mergeStickyArgsFromChildView` method and allow the parent component to indicate that the child view args should propagate (from 1st level child or even from deeped level).

I belive when `mergeStickyArgsFromChildView` is kept, it can help solving https://github.com/atk4/ui/pull/1649#discussion_r672293568

Reply:
> If Callback has some args that should be sticky, with `mergeStickyArgsFromChildView` the args are stored inside the Callback object. If that object is mutated to some different state, the parent/independent view CAN (but does not have to) reflect that.

Only the View consuming the callback should have the stickyArgs, not the Callback itself.

Reply:
https://github.com/atk4/ui/pull/1646/files#diff-489a31f422a2238af13855d2db35e0df62c99606b21cb5c678330f7f857ea90eR43-R47

really? This method sets an URL trigger and makes it sticky. Sticky to the parent, thus when any URL from parent is generated, it is always sticky, even if Callback trigger is not wanted.

Example render tree:

```
App - View (alias V1) - Callback 1 (alias C1)
                      - Callback 2 (alias C2)
```

Let's imagine C1 is triggered.

C1 is made sticky. Can we agree here?

Now, if C1 wants to render a V1 or C2 URL, it can not, as V1 contains sticky args from C1. This is a problem. Can we agree here too?

Reply:
View generates proper Url for the Callbacks - It should know which one to generate. Callback is not a View and it is not rendered. Only the arguments that need to fire the callback should be carry by the View and this should be done by the View itself when generating the URL.

Reply:
> Callback is not a View and it is not rendered.

It is rendered, eg. triggered = executed = rendered but terminated with a custom output, right?

What is the diffrent vs. VirtualPage?

The situation I describe is even more evident with this render tree:

```
App - View 1 (alias V1) - Callback 1 (alias C1)
                        - View 2 (alias V2)
```

When C1 is triggered and C1 is supposed (because coder wants) to return V2 URL, V2 will contains sticky args from C1, which is wrong.

Reply:
> It is rendered, eg. triggered = executed = rendered but terminated with a custom output, right?

Callback itself is not rendered. A View is, depending on the implementation.  There are two arguments need to run the callback function. One argument will trigger the callback method (canTriggered) and the other will output content via the terminate method (canTerminate).

 These arguments are:
 - The triggering argument:
    - The callback name (default) = the callback mode (mode is used in VirtualPage and determine the type of output)
       When the View consuming the callback reaches the triggering argument, then the callback method is executed.
- The current callback or callback that can terminate output.
  - __atk_callback = the callback name (default)
    This is the current callback that can terminate app execution.

Let's take VirtualPage demo https://github.com/atk4/ui/blob/develop/demos/_unit-test/virtual-page.php as an example.

In order to reach the callback that populates the Dropdown control, all previous VirtualPage callback method has to be triggered thus carrying all triggered callback argument with the URL firing the dropdown.

See below the URL firing the Dropdown request for content. As you can see from the picture, only one callback can terminate output, the Dropdown control in this case.

<img width="1121" alt="Screen Shot 2021-07-19 at 5 46 33 PM" src="https://user-images.githubusercontent.com/2204478/126231669-a83e5c67-b854-4261-9364-4d1c8ba270ae.png">

If you missed one triggering argument, then the Dropdown callback can never be reached.

Reply:
> In order to reach the callback that populates the Dropdown control, all previous VirtualPage callback method has to be triggered thus carrying all triggered callback argument with the URL firing the dropdown.

This is clear to me. VP is a component that can be nested, Callback is the possible execution target (eg. can not be nested).

> When C1 is triggered and C1 is supposed (because coder wants) to return V2 URL, V2 will contains sticky args from C1, which is wrong.

`__atk_callback` will not be set, thus it will behave correctly, but is there any need for sticky V1 or C1 args if V2 is rendered from triggered C1?

Reply:
> Is there any reason to use Callback over VP for the end-user?

Loader, Form, Console; they all use a callback. Loader can be nested. They can even be nested inside a VirtualPage.

I do not see any problem carrying over argument. A callback is used to terminate output at some point. Before the regular app->output happen. Thus you want to make sure that View added within View that consume the callback can triggered this one.

Reply:
> Loader, Form, Console; they all use a callback. Loader can be nested. They can even be nested inside a VirtualPage.

but nested using View class...

> I do not see any problem carrying over argument. A callback is used to terminate output at some point. Before the regular app->output happen. Thus you want to make sure that View added within View that consume the callback can triggered this one.

URL of a View should never be affected by a component that is NOT within a sub render tree nor parents. This is important for debugging or for things like immutable state/diffing. So V2 URL must be rendered always the same no matter if rendered from App/V1 or C1.

To achive this, C1 sticky args must never be saved within V1. This usecase should be also covered by a test.

Reply:
> To achive this, C1 sticky args must never be saved within V1. This usecase should be also covered by a test.

Done! Also had use case for it.

Reply:
Can the 2nd arg be an instance of the running Callback?

If yes, it will be more robust as we will not have to filter `$_GET` which can be spoofed. We can easily traverse the passed Callback to it's root/parents/App and collect all triggering names.